### DISCUSSION TITLE
GCC supports ISA extenstion split `A` into `zaamo` and `zalrsc`

### DISCUSSION THREAD

**Discussion by NikLeberg** [stnolting/neorv32#1051] [General]
Hey there @stnolting 

Quick heads-up: I found that GCC finally supports emit of lr/sc sequences for the builtin atomic operations with setting `-march=<as usual>_zalrsc`.

https://godbolt.org/z/j39MoeeKv

I just love how risc-v keeps on getting better bit-by-bit. 😍

**Comment by stnolting** [comment]
Oh that's nice! Now I really need to check if newlib also supports `stdatomic.h`... then we could remove this manual AMO emulation library. 🤔 

I wonder if GCC also supports built-in defines for that so that we could concretize those `__riscv_atomic` ifdefs:

https://github.com/stnolting/neorv32/blob/f1069d114f73521e44a563a34b651e405a6f086a/sw/lib/include/neorv32_cpu_amo.h#L31-L45

> I just love how risc-v keeps on getting better bit-by-bit. 😍

Absolutely! ❤️

**Reply by NikLeberg** [reply]
If newlib does not support `stdatomic.h` one could also just use the _actual_ atomic operations that GCC uses internally to implement the _stdatomic_ ones. 
https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html

**Reply by stnolting** [reply]
I've never tried them either, but it's worth a try!
